Домашнее задание 2 - 10 баллов
В этом задании вам предстоит продолжить работу с датасетом lenta-ru-news для той же задачи - классификации текстов по топикам. Можно переиспользовать подготовленные данные из ДЗ 1 или загрузить их заново.


In [ ]:
!pip install gensim

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 64.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 16.9 MB/s eta 0:00:00
  Attempting uninstall: scipy
    Found existing installation: scipy 1.14.1
    Uninstalling scipy-1.14.1:
      Successfully uninstalled scipy-1.14.1


In [ ]:
import pandas as pd
import numpy as np
import re
import random
from tqdm import tqdm
import string

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer

import gensim
from gensim.models import Word2Vec, KeyedVectors


In [ ]:
!pip install navec > /dev/null

In [ ]:
from navec import Navec

RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
random.seed(RANDOM_SEED)


# 1. Разделите датасет на обучающую, валидационную и тестовую выборки со стратификацией в пропорции 60/20/20. В качестве целевой переменной используйте атрибут topic

In [ ]:
!pip install corus

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.7/83.7 kB 4.2 MB/s eta 0:00:00


In [ ]:
!wget https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz

--2025-03-14 08:31:18--  https://github.com/yutkin/Lenta.Ru-News-Dataset/releases/download/v1.0/lenta-ru-news.csv.gz
Resolving github.com (github.com)... 140.82.116.4
Connecting to github.com (github.com)|140.82.116.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250314%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250314T083119Z&X-Amz-Expires=300&X-Amz-Signature=464157094e0af2d79d956cefddd2605c4f445d171e0a253827f7483f0ea62880&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dlenta-ru-news.csv.gz&response-content-type=application%2Foctet-stream [following]
--2025-03-14 08:31:19--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/87156914/0b363e00-0126-11e9-9e3c-e8c235463bd6?X-Amz-Algorithm=AWS4-HMAC-

In [ ]:
from corus import load_lenta

path = 'lenta-ru-news.csv.gz'
records = load_lenta(path)

data = []
MAX_SIZE = 200_000  # чтобы не было слишком долго
for i, r in enumerate(records):
    if r.text is not None and r.topic is not None:
        data.append({
            'text': r.text,
            'topic': r.topic
        })
    if len(data) >= MAX_SIZE:
        break

df = pd.DataFrame(data)
df.dropna(subset=['text', 'topic'], inplace=True)
df.reset_index(drop=True, inplace=True)
df.shape


(200000, 2)

In [ ]:
df["topic"], uniques = pd.factorize(df["topic"])
print("Всего уникальных тем:", len(uniques))

train_val, test = train_test_split(
    df,
    test_size=0.2,
    stratify=df["topic"],
    random_state=RANDOM_SEED
)
train, val = train_test_split(
    train_val,
    test_size=0.25,  # 25% от 80% = 20% от всего => итого 60/20/20
    stratify=train_val["topic"],
    random_state=RANDOM_SEED
)

X_train, y_train = train["text"].values, train["topic"].values
X_val,   y_val   = val["text"].values,   val["topic"].values
X_test,  y_test  = test["text"].values,  test["topic"].values

print("Train:", X_train.shape, "Val:", X_val.shape, "Test:", X_test.shape)


Всего уникальных тем: 20
Train: (120000,) Val: (40000,) Test: (40000,)


# 2. Обучите word2vec-эмбеддинги с помощью библиотеки gensim - 2 балла
* создайте модель для обучения на ваших данных, опишите, какими значениями вы
инициализировали гиперпараметры модели, и почему
* визуально оцените внутреннее (intrinsic) качество получившихся эмбеддингов, используя методы gensim - doesnt_match, most_similar

In [ ]:
def simple_preprocess(text):
    text = text.lower()
    text = re.sub(r'[^а-яёa-z\s]', ' ', text)
    tokens = text.split()
    return tokens

# Для ускорения возьмём меньше данных для обучения эмбеддингов
w2v_texts = train["text"].sample(50_000, random_state=RANDOM_SEED, replace=False).values

sentences = []
for t in tqdm(w2v_texts):
    tokens = simple_preprocess(t)
    if tokens:
        sentences.append(tokens)

print(f"Кол-во предложений: {len(sentences)}")
print(f"Пример: {sentences[0][:20]}")


100%|██████████| 50000/50000 [00:06<00:00, 7917.29it/s]

Кол-во предложений: 49999
Пример: ['граждане', 'германии', 'и', 'других', 'стран', 'все', 'еще', 'хранят', 'миллиарда', 'немецких', 'марок', 'хотя', 'с', 'момента', 'полного', 'перехода', 'на', 'евро', 'прошло', 'почти']


In [ ]:
from gensim.models import Word2Vec

w2v_model = Word2Vec(
    sentences=sentences,
    vector_size=300,  # размерность
    window=5,         # размер контекстного окна
    min_count=5,      # игнорируем слова реже 5
    sg=1,             # 1 = skip-gram, 0 = cbow
    negative=5,       # negative sampling
    epochs=10,        # кол-во эпох
    seed=RANDOM_SEED,
    workers=4         # кол-во потоков
)


print(w2v_model.wv['россия'][:10])


[ 0.14236239 -0.03583811 -0.05071652  0.0233071  -0.8749301   0.45734003
  0.00998669  0.38948664 -0.33318004  0.5485556 ]


In [ ]:
w2v_model.wv.most_similar('россия', topn=5)

[('украина', 0.547884464263916),
 ('белоруссия', 0.5341498851776123),
 ('турция', 0.4955958425998688),
 ('франция', 0.46975839138031006),
 ('страна', 0.4617319703102112)]

In [ ]:
w2v_model.wv.doesnt_match(['зима','лето','апрель','красный'])

'красный'

# 3. Загрузите предобученные эмбеддинги из navec и rusvectores (на ваш вкус) - 1 балл

In [ ]:
!wget -q https://storage.yandexcloud.net/natasha-navec/packs/navec_news_v1_1B_250K_300d_100q.tar

In [ ]:
!pip install navec

In [ ]:

from navec import Navec

navec_path = 'navec_news_v1_1B_250K_300d_100q.tar'
navec_model = Navec.load(navec_path)

print(len(navec_model.vocab.words))
print(navec_model['россия'][:10])


250002
[-0.18471098 -0.11100645  0.08644193  0.2270673   0.02942725  0.3612441
  0.6843062   0.30184293  0.5982135   0.32763943]


In [ ]:
!wget -q  https://rusvectores.org/static/models/rusvectores4/ruwikiruscorpora/ruwikiruscorpora_upos_skipgram_300_2_2018.vec.gz

In [ ]:
import gensim
rusvectores_path = 'ruwikiruscorpora_upos_skipgram_300_2_2018.vec.gz'
rusvectores_model = gensim.models.KeyedVectors.load_word2vec_format(rusvectores_path)


# 4. Обучите модель sklearn.linear_model.LogisticRegression с тремя вариантами векторизации текстов и сравните их качество между собой на валидационной выборке: 2 балла
* ваши эмбеддинги w2v
* предобученные эмбеддинги navec
* предобученные эмбеддинги rusvectores

In [ ]:
def text_to_vec_mean(text, emb_model, dim=300):
    tokens = simple_preprocess(text)
    vecs = []
    for t in tokens:
        if t in emb_model:
            vecs.append(emb_model[t])
    if len(vecs) == 0:
        return np.zeros(dim)
    else:
        return np.mean(vecs, axis=0)

vec = text_to_vec_mean("Пример текста о России", w2v_model.wv, dim=300)
print(vec.shape)


(300,)


In [ ]:
def transform_texts_to_matrix(texts, emb_model, dim=300):
    X_matrix = np.zeros((len(texts), dim), dtype='float32')
    for i, doc in enumerate(texts):
        X_matrix[i] = text_to_vec_mean(doc, emb_model, dim)
    return X_matrix


In [ ]:
dim_w2v = 300
X_train_w2v = transform_texts_to_matrix(X_train, w2v_model.wv, dim=dim_w2v)
X_val_w2v   = transform_texts_to_matrix(X_val,   w2v_model.wv, dim=dim_w2v)


In [ ]:
dim_navec = 300
X_train_navec = transform_texts_to_matrix(X_train, navec_model, dim=dim_navec)
X_val_navec   = transform_texts_to_matrix(X_val,   navec_model, dim=dim_navec)


In [ ]:
clf_w2v = LogisticRegression(random_state=RANDOM_SEED, max_iter=1000)
clf_w2v.fit(X_train_w2v, y_train)
val_pred_w2v = clf_w2v.predict(X_val_w2v)
acc_w2v = accuracy_score(y_val, val_pred_w2v)
print("Accuracy (w2v) on val:", acc_w2v)


Accuracy (w2v) on val: 0.792325


In [ ]:
clf_navec = LogisticRegression(random_state=RANDOM_SEED, max_iter=1000)
clf_navec.fit(X_train_navec, y_train)
val_pred_navec = clf_navec.predict(X_val_navec)
acc_navec = accuracy_score(y_val, val_pred_navec)
print("Accuracy (navec) on val:", acc_navec)


Accuracy (navec) on val: 0.7934


In [ ]:
!pip install pymorphy3

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.8/53.8 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.4/8.4 MB 74.7 MB/s eta 0:00:00


In [ ]:
# import pymorphy3

# morph = pymorphy3.MorphAnalyzer()

# def tokenize_with_pos(text):
#     # разбиение текста по пробелам или через regex
#     raw_tokens = text.split()
#     result = []
#     for raw in raw_tokens:
#         parsed = morph.parse(raw)[0]
#         lemma = parsed.normal_form
#         pos = parsed.tag.POS  # например, NOUN/VERB/ADJ
#         if pos is not None:
#             token_pos = f"{lemma}_{pos}"
#             result.append(token_pos)
#         else:
#             # если POS не определено, можно пропустить или оставить как есть
#             pass
#     return result

# def transform_texts_to_matrix_rusv(texts, rusvectores_model, dim=300):
#     X_matrix = np.zeros((len(texts), dim), dtype='float32')
#     for i, text in enumerate(texts):
#         tokens = tokenize_with_pos(text)
#         vectors = []
#         for tok in tokens:
#             if tok in rusvectores_model:
#                 vectors.append(rusvectores_model[tok])
#         if vectors:
#             X_matrix[i] = np.mean(vectors, axis=0)
#         else:
#             X_matrix[i] = np.zeros(dim)
#     return X_matrix
#
# dim_rusvectores = 300
# X_train_rusv = transform_texts_to_matrix_rusv(X_train, rusvectores_model, dim=dim_rusvectores)
# X_val_rusv   = transform_texts_to_matrix_rusv(X_val,   rusvectores_model, dim=dim_rusvectores)
#
# clf_rusv = LogisticRegression(random_state=RANDOM_SEED, max_iter=200)
# clf_rusv.fit(X_train_rusv, y_train)
# val_pred_rusv = clf_rusv.predict(X_val_rusv)
# acc_rusv = accuracy_score(y_val, val_pred_rusv)
# print("Accuracy (rusvectores) on val:", acc_rusv)


In [ ]:
import numpy as np
import re
import pymorphy3
from functools import lru_cache
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

morph = pymorphy3.MorphAnalyzer()

@lru_cache(maxsize=100_000)
def parse_cached(word):
    return morph.parse(word)[0]

def tokenize_with_pos(text, limit_tokens=300):

    # обрежем текст до первых 300 слов, чтобы сильно ускорить:
    raw_tokens = text.split()[:limit_tokens]

    result = []
    for raw in raw_tokens:
        raw_lower = raw.lower()
        parsed = parse_cached(raw_lower)
        lemma = parsed.normal_form
        pos = parsed.tag.POS  # например, NOUN/VERB/ADJ
        if pos is not None:
            token_pos = f"{lemma}_{pos}"
            result.append(token_pos)
    return result

def transform_texts_to_matrix_rusv(texts, rusvectores_model, dim=300):
    X_matrix = np.zeros((len(texts), dim), dtype='float32')
    for i, text in enumerate(texts):
        tokens = tokenize_with_pos(text)
        vectors = []
        for tok in tokens:
            if tok in rusvectores_model:
                vectors.append(rusvectores_model[tok])
        if vectors:
            X_matrix[i] = np.mean(vectors, axis=0)
        else:
            # Если ни одного слова нет в модели, ставим нулевой вектор
            X_matrix[i] = np.zeros(dim, dtype='float32')
    return X_matrix

dim_rusvectores = 300
X_train_rusv = transform_texts_to_matrix_rusv(X_train, rusvectores_model, dim=dim_rusvectores)
X_val_rusv   = transform_texts_to_matrix_rusv(X_val,   rusvectores_model, dim=dim_rusvectores)



In [ ]:

clf_rusv = LogisticRegression(random_state=RANDOM_SEED, max_iter=1000)
clf_rusv.fit(X_train_rusv, y_train)


LogisticRegression(max_iter=1000, random_state=42)

In [ ]:

val_pred_rusv = clf_rusv.predict(X_val_rusv)
acc_rusv = accuracy_score(y_val, val_pred_rusv)
print("Accuracy (rusvectores) on val:", acc_rusv)

Accuracy (rusvectores) on val: 0.7225


# 5. Попробуйте улучшить качество модели, взяв для ее обучения лучший набор эмбеддингов и используя его с взвешиванием через tf-idf. То есть, необходимо каждый текст представить в виде взвешенного усреднения эмбеддингов его слов, где весами являются соответствующие коэффициенты tf-idf - 2 балла

In [ ]:
tfidf = TfidfVectorizer(
    tokenizer=simple_preprocess,
    max_df=0.7,
    min_df=5
)
tfidf.fit(X_train)

# Получим словарь (token->index) + idf
vocab = tfidf.vocabulary_
idf = tfidf.idf_


/usr/local/lib/python3.11/dist-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [ ]:
def text_to_vec_tfidf(text, emb_model, vocab, idf_values, dim=300):
    tokens = simple_preprocess(text)
    # Подсчитаем частоту токенов
    counts = {}
    for t in tokens:
        if t not in counts:
            counts[t] = 0
        counts[t] += 1
    # Общая длина
    total = len(tokens)

    vec_sum = np.zeros(dim, dtype='float32')
    weight_sum = 0.0

    for t, freq in counts.items():
        if t in vocab and t in emb_model:
            tf = freq / total
            idx = vocab[t]  # индекс токена в словаре tfidf
            # idf = idf_values[idx]
            idf_t = idf_values[idx]

            w = tf * idf_t
            vec_sum += emb_model[t] * w
            weight_sum += w

    if weight_sum == 0:
        return vec_sum
    else:
        return vec_sum / weight_sum


In [ ]:
X_train_navec_tfidf = np.zeros((len(X_train), 300), dtype='float32')
for i, doc in enumerate(X_train):
    X_train_navec_tfidf[i] = text_to_vec_tfidf(doc, navec_model, vocab, idf, dim=300)

X_val_navec_tfidf = np.zeros((len(X_val), 300), dtype='float32')
for i, doc in enumerate(X_val):
    X_val_navec_tfidf[i] = text_to_vec_tfidf(doc, navec_model, vocab, idf, dim=300)


In [ ]:
clf_navec_tfidf = LogisticRegression(random_state=RANDOM_SEED, max_iter=200)
clf_navec_tfidf.fit(X_train_navec_tfidf, y_train)

val_pred_navec_tfidf = clf_navec_tfidf.predict(X_val_navec_tfidf)
acc_navec_tfidf = accuracy_score(y_val, val_pred_navec_tfidf)
print("Accuracy (navec + tf-idf weighting) on val:", acc_navec_tfidf)


Accuracy (navec + tf-idf weighting) on val: 0.790975


# 6. Финально сравните качество всех моделей на тестовой выборке - 1 балл

In [ ]:
# 1) Собственный W2V
X_test_w2v = transform_texts_to_matrix(X_test, w2v_model.wv, 300)
test_pred_w2v = clf_w2v.predict(X_test_w2v)
acc_test_w2v = accuracy_score(y_test, test_pred_w2v)

print("Test accuracy (W2V) =", acc_test_w2v)

Test accuracy (W2V) = 0.791625


In [ ]:
# 2) Navec
X_test_navec = transform_texts_to_matrix(X_test, navec_model, 300)
test_pred_navec = clf_navec.predict(X_test_navec)
acc_test_navec = accuracy_score(y_test, test_pred_navec)

print("Test accuracy (Navec) =", acc_test_navec)

Test accuracy (Navec) = 0.79235


In [ ]:
# 3) RusVectores
dim_rusvectores = 300
X_test_rusv = transform_texts_to_matrix_rusv(X_test, rusvectores_model, dim=dim_rusvectores)

test_pred_rusv = clf_rusv.predict(X_test_rusv)
acc_test_rusv = accuracy_score(y_test, test_pred_rusv)

print("Test accuracy (RusVectores) =", acc_test_rusv)

Test accuracy (RusVectores) = 0.7242


In [ ]:
# 4) tf-idf weighting + navec
X_test_navec_tfidf = np.zeros((len(X_test), 300), dtype='float32')
for i, doc in enumerate(X_test):
    X_test_navec_tfidf[i] = text_to_vec_tfidf(doc, navec_model, vocab, idf, 300)
test_pred_navec_tfidf = clf_navec_tfidf.predict(X_test_navec_tfidf)
acc_test_navec_tfidf = accuracy_score(y_test, test_pred_navec_tfidf)

print("Test accuracy (Navec + tfidf) =", acc_test_navec_tfidf)

Test accuracy (Navec + tfidf) = 0.7896


# Результаты

Мы выбрали собственный Word2Vec и сравнили его с предобученными моделями Navec  и RusVectores. Для классификации взяли LogisticRegression, так как она быстро обучается на эмбеддингах размерности 300 и позволяет легко сравнить разные векторизации.

Усреднение эмбеддингов выбрано как простой способ получить вектор текста, а TF-IDF-веса были добавлены, чтобы усилить вклад редких, информативных слов и ослабить слишком частотные. Все гиперпараметры (размер окна, размер векторов, min_count, тип Skip-Gram) подобраны как компромисс между качеством и скоростью обучения(стандартные удачные).

В итоге лучший результат показал Navec (около 0.79), за ним с небольшим отрывом идёт собственный Word2Vec (0.79), тогда как RusVectores остановился на 0.72. Добавление TF-IDF-весов в данном случае не улучшило итоговую метрику. Таким образом, Navec оказался наиболее эффективным вариантом для классификации новостей по темам в этом датасете.